File: vol1/ch2/LeetCode23-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0) return NULL;
        ListNode*h=lists[0];
        for(int i=1;i<lists.size();i++) {
            h=merge(h,lists[i]);
        }
        return h;
    }
    ListNode*merge(ListNode* list1,ListNode* list2) {
        if(list1==NULL) return list2;
        if(list2==NULL) return list1;
        ListNode*h=new ListNode(-1),*r=h;
        ListNode*p=list1,*q=list2;
        while(p!=NULL && q!=NULL) {
            if(p->val<q->val) {
                r->next=p; r=p;
                p=p->next;
            }
            else {
                r->next=q; r=q;
                q=q->next;
            }
        }
        if(p!=NULL) r->next=p;
        else r->next=q;
        return h->next;
    }
};

File: vol1/ch2/LeetCode23-1.py
class Solution:
    def mergeKLists(self,lists:List[Optional[ListNode]])->Optional[ListNode]:
        k=len(lists)
        if k==0:return None
        h=lists[0]
        for i in range(1,k):
            h=self.merge(h,lists[i])
        return h

    def merge(self,list1,list2):
        if list1==None: return list2
        if list2==None: return list1
        h=ListNode(-1); r=h
        p=list1; q=list2
        while p!=None and q!=None:
            if p.val<q.val:
                r.next=p; r=p
                p=p.next
            else:
                r.next=q; r=q
                q=q.next
        if p!=None: r.next=p
        else: r.next=q
        return h.next
